web_client/actions.jsx

```diff
@@ -223,6 +223,101 @@ export const enableAnalysisPage = (page) => promiseAction(
   }
 );
 
+export const ensurePrivateDirectory = (() => {
+  let singleton;
+
+  return (args) => {
+    if (!singleton) {
+      singleton = ensureUserDirectory({
+        name: 'Private',
+        description: 'Private Directory',
+        ...args
+      });
+    }
+
+    return singleton;
+  };
+})();
+
+export const ensureScratchDirectory = (() => {
+  let singleton;
+
+  return (args) => {
+    if (!singleton) {
+      singleton = ensureUserDirectory({
+        name: '__osumo_tmp',
+        description: 'Temporary Staging Space',
+        ...args
+      });
+    }
+
+    return singleton;
+  };
+})();
+
+export const ensureUserDirectory = (args) => {
+  let {
+    description,
+    name,
+    returnModel,
+    user
+  } = args;
+
+  return promiseAction(
+    (dispatch, getState) => {
+      if (isNil(returnModel)) {
+        returnModel = true;
+      }
+
+      if (isNil(user)) {
+        ({ user } = getState().loginInfo);
```

Comment:
Needed in this case to disambiguate syntax from that of a new code block.

i.e.:

    { user } = ...;

 fails to parse, as it looks like a code block followed by an `=` operator.